### ISSUE TITLE
Allow output to file with different format to output to console

### ISSUE THREAD

**Issue by ngnpope** [astral-sh/ruff#2388]
Currently, when running in CI I have to do the following, e.g. in a (simplified) `.gitlab-ci.yml`:

```yaml
lint:
  script:
    - ...
    - ruff .
    - ruff --format=gitlab . > quality.json
  artifacts:
    reports:
      codequality: quality.json
```

I'd like to keep the text output in the job output, but also make use of the capabilities GitLab provides to display violations.

Now, `ruff` is super fast ⚡, so running twice isn't a pain, but it seems suboptimal. Could we enable one of the following?

- `ruff --format=text --output=gitlab:/path/to/quality.json .` ← dedicated `--output` option
- `ruff --format=text --format=gitlab:/path/to/quality.json .` ← allow multiple `--format` with filename
- `ruff --format=text,gitlab:/path/to/quality.json .` ← allow multiple values in `--format`

**Comment by not-my-profile** [comment]
I'd rather avoid CLI options with such a complex syntax, since it complicates the CLI and also doesn't work with the shell autocompletion generation by `clap_complete`.

How about instead introducing a new `--stderr-format` option that accepts the same values as the current `--format` option and would additionally output the specified format on stderr? ... then you could replace your two commands with `ruff --stderr-format=gitlab . 2> quality.json` (`2>` is shell syntax for stderr redirection).

**Comment by ngnpope** [comment]
I guess we could do that, but that would only work as long as `ruff` never chose to output something else to stderr...

It seems a little irregular. Normally stdout is used for the normal output from a program and stderr is use to output errors or diagnostic information about the execution of that program.

**Comment by not-my-profile** [comment]
Fair point ... and it also wouldn't support outputting to more than two formats. I think it's probably best to just implement a new `convert` subcommand that can consume the `json` output and convert it to any of the other supported output formats.

**Comment by ngnpope** [comment]
So `--format=jsonl` then? (See https://jsonlines.org/)

I guess that allows for using `tee` to redirect to multiple processors then?

It would work, but does just seem much more complex than a command line option... 🤔

**Comment by not-my-profile** [comment]
The advantages I see are that 1) it wouldn't complicate the CLI for users who don't need this feature, 2) shell completion would work fine and 3) it's not unthinkable that we end up introducing more format-specific options (for `--format text` we already have several), having one ruff command call always only produce one output avoids the ambiguity of which output option belongs to which output.

I agree that how to use this would be less obvious for users unfamiliar with shell piping/redirection, however I think that could be accommodated via an example in the documentation ... and I wouldn't consider this to be something users commonly use but rather something you set up once for a CI, so I think that should be alright.

**Comment by MichaReiser** [comment]
Having a dedicated `output` (or `output-file`) option has precedence in 

* ESLint (`output-file` | `-o`)
* `dotnet format --report`
* Jest (`--outputFile`) in combination with `--json`

**Comment by ngnpope** [comment]
I should follow up that, while the PR you linked to (#4950) does support writing to an output file with a flag, it doesn't do anything other than remove the need for shell redirection.

**Comment by MichaReiser** [comment]
CC: @dhruvmanila

**Comment by mgorfer** [comment]
I would also be very glad to have this feature included in Ruff (https://github.com/astral-sh/ruff/discussions/6467).

**Comment by ekbfh** [comment]
@charliermarsh Good day! Is any new thoughts about CI usage of ruff? We stiil have to run it twice in a row

**Comment by gmetzker-4c** [comment]
Agreed this would be a nice feature.  
Pylint now supports multiple output formats as follows:  
`pylint --output-format=colorized,parseable:pylint.log`

That's not to say their syntax is the best but the ability target multiple outputs would be great for CI systems.

**Comment by MarcBresson** [comment]
Hello, any news on that issue ?

**Comment by wkoot** [comment]
I would like to use ruff to replace pylint in a pipeline, but having to run several times seems a bit of a shame.
The sonar-scanner provides the option to pass a pylint report via `sonar.python.pylint.reportPaths`, which ruff can produce.
However, this would mean I'd need to run 3 times; one for immediate readable output, one for gitlab annotation and one for pylint.

I think that the use case is rather specific so I don't think making the cli overly complicated is needed, toml config should suffice.
For instance, something like this:
```toml
[tool.ruff]
output-format = "concise"

[tool.ruff.per-file-output]
ruff_gitlab.json = {output-format = "gitlab"}
ruff_junit.xml = {output-format = "junit"}
ruff_pylint.txt = {output-format = "pylint"}
```
